### PR TITLE
Reconcile costfuncs with pars

### PR DESCRIPTION
**Pull request by cliffckerr**

### CODE REVIEW

File: optima/programs.py [hunk 1]
```diff
@@ -578,10 +600,35 @@ def compareoutcomes(self, parset=None, year=None, ind=0, doprint=False):
     
     
     
-    def reconcilewithpars(self, parset=None, year=None, ind=0):
+    def cco2odict(self, t=None):
+        ''' Parse the cost-coverage-outcome tree and pull out parameter values into an odict '''
+        if t is None: raise Exception('Please supply a year')
```
Comment (robynstuart) [inline optima/programs.py]: i refer you to your previous comment re. OptimaExceptions

Comment (robynstuart) [inline optima/programs.py]: oh, except maybe i wrote this bit

File: optima/programs.py [hunk 2]
```diff
@@ -578,10 +600,35 @@ def compareoutcomes(self, parset=None, year=None, ind=0, doprint=False):
...
+        modifiablepars = odict()
+        for targetpartype in self.covout.keys():
+            for targetparpop in self.covout[targetpartype].keys():
+                modifiablepars[(targetpartype,targetparpop,'intercept')] = [self.covout[targetpartype][targetparpop].getccopar(t=t,bounds='l')['intercept'][0],self.covout[targetpartype][targetparpop].getccopar(t=t,bounds='u')['intercept'][0]]
+                for thisprog in self.progs_by_targetpar(targetpartype)[targetparpop]:
+                    try: modifiablepars[(targetpartype,targetparpop,thisprog.short)] = [self.covout[targetpartype][targetparpop].getccopar(t=t,bounds='lower')[thisprog.short][0], self.covout[targetpartype][targetparpop].getccopar(t=t,bounds='upper')[thisprog.short][0]]
+                    except: pass # Must be something like ART, which does not have adjustable parameters -- WARNING, could test explicitly!
```
Comment (robynstuart) [inline optima/programs.py]: this is going to be much easier to handles later on, when i change programs to use Par objects directly.

File: optima/programs.py [hunk 3]
```diff
@@ -578,10 +600,35 @@ def compareoutcomes(self, parset=None, year=None, ind=0, doprint=False):
...
+        return modifiablepars
+
+
+
+    def odict2cco(self, modifiablepars=None, t=None):
+        ''' Take an odict and use it to update the cost-coverage-outcome tree '''
+        if modifiablepars is None: raise Exception('Please supply modifiablepars')
+        for key,val in modifiablepars.items():
+            targetpartype,targetparpop,thisprogkey = key # Split out tuple
+            self.covout['condcas'][('Clients', 'FSW')].ccopars[thisprogkey] = [tuple(val)]
```
Comment (robynstuart) [inline optima/programs.py]: this doesn't look like it should be here...

File: optima/programs.py [hunk 4]
```diff
@@ -578,10 +600,35 @@ def compareoutcomes(self, parset=None, year=None, ind=0, doprint=False):
...
+            if t: self.covout['condcas'][('Clients', 'FSW')].ccopars['t'] = t # WARNING, reassigned multiple times, but shouldn't matter...
```
Comment (robynstuart) [inline optima/programs.py]: nor this.

Comment (robynstuart) [inline optima/programs.py]: NB, i can fix this if you like

File: optima/programs.py [hunk 5]
```diff
@@ -603,29 +650,47 @@ def objectivecalc(progset=None, parset=None, year=None, ind=None, method='mape',
                 mismatch += thismismatch
             return mismatch
         
-        ## Do the actual calibration thingo
-        def cco2vec(): raise Exception('Not implemented')
+        # Store original values in case we need to go back to them
+        origcovout = dcp(self.covout)
+        origvals = dcp(self.cco2odict(t=year))
+        workingvals = dcp(origvals)
+        testarr = origvals[:] # Turn into array format
```
Comment (robynstuart) [inline optima/programs.py]: is this the bit that's a 2xN vector when you want a 1x2N vector? should be easy to fix if so

**Comment by cliffckerr** [comment]
doesn't crash, but doesn't work yet

**Pull request closed without merging**